nat-lab/tests/test_direct_connection.py

```diff
@@ -39,265 +40,227 @@
+    ],
+    [
+        (ConnectionTag.DOCKER_SYMMETRIC_CLIENT_1, ["stun"], True, DOCKER_SYMMETRIC_GW_1_IP),
+        (ConnectionTag.DOCKER_FULLCONE_CLIENT_1, ["stun"], True, DOCKER_FULLCONE_GW_1_IP),
+        (ConnectionTag.DOCKER_FULLCONE_CLIENT_2, ["stun"], False, DOCKER_FULLCONE_GW_2_IP),
+        (ConnectionTag.DOCKER_OPEN_INTERNET_CLIENT_1, ["stun"], True, DOCKER_OPEN_INTERNET_CLIENT_1_IP),
+        (ConnectionTag.DOCKER_OPEN_INTERNET_CLIENT_2, ["stun"], False, DOCKER_OPEN_INTERNET_CLIENT_2_IP),
+        (ConnectionTag.DOCKER_OPEN_INTERNET_CLIENT_DUAL_STACK, ["stun"], True, DOCKER_OPEN_INTERNET_CLIENT_DUAL_STACK_IP),
+    ],
+    [
+        (ConnectionTag.DOCKER_OPEN_INTERNET_CLIENT_1, ["local"], True, DOCKER_OPEN_INTERNET_CLIENT_1_IP),
+        (ConnectionTag.DOCKER_SYMMETRIC_CLIENT_1, ["local"], True, DOCKER_OPEN_INTERNET_CLIENT_1_IP),
+    ],
+    [
+        (ConnectionTag.DOCKER_SYMMETRIC_CLIENT_1, ["local"], True, DOCKER_SYMMETRIC_CLIENT_1_IP),
+        (ConnectionTag.DOCKER_INTERNAL_SYMMETRIC_CLIENT, ["local"], True, DOCKER_SYMMETRIC_CLIENT_1_IP),
+    ],
 ]
+# fmt: on
 
 
-@pytest.mark.asyncio
-@pytest.mark.long
-@pytest.mark.timeout(timeouts.TEST_DIRECT_FAILING_PATHS_TIMEOUT)
-@pytest.mark.parametrize("setup_params", UHP_FAILING_PATHS)
-# Not sure this is needed. It will only be helpful to catch if any
-# libtelio change would make any of these setup work.
-async def test_direct_failing_paths(setup_params: List[SetupParameters]) -> None:
-    async with AsyncExitStack() as exit_stack:
-        env = await setup_mesh_nodes(exit_stack, setup_params, is_timeout_expected=True)
-        _, beta = env.nodes
-        alpha_client, beta_client = env.clients
-        alpha_connection, _ = [conn.connection for conn in env.connections]
+async def _ping_between_all_nodes(env: Environment) -> None:
+    await asyncio.gather(*[
+        ping(conn.connection, node.ip_addresses[0])
+        for (client, conn), node in itertools.product(
+            zip(env.clients, env.connections), env.nodes
+        )
+        if not client.is_node(node)
+    ])
 
-        for server in DERP_SERVERS:
-            await exit_stack.enter_async_context(
-                alpha_client.get_router().break_tcp_conn_to_host(str(server["ipv4"]))
+
+async def _check_if_true_direct_connection(env: Environment) -> None:
+    async with AsyncExitStack() as temp_exit_stack:
+        await asyncio.gather(*[
+            temp_exit_stack.enter_async_context(
+                client.get_router().break_tcp_conn_to_host(str(server["ipv4"]))
             )
-            await exit_stack.enter_async_context(
-                beta_client.get_router().break_tcp_conn_to_host(str(server["ipv4"]))
+            for client, server in itertools.product(env.clients, DERP_SERVERS)
+        ])
+
+        await asyncio.gather(*[
+            client.wait_for_state_on_any_derp([State.Connecting, State.Disconnected])
+            for client in env.clients
+        ])
```

Comment:
Prepare wait futures before breaking the connection. Same for State.Connected. Should be created before restoring the connection

Reply:
Since you are checking for state, you can do that after the fact. Doesn't really matter if future is created before imo. If you were to check for event, then yeah, waiting for event beforehand would make sense.

Reply:
Ohh, I was not aware of that distinction.  I checked the code now, and you're right. Thanks!